File: apple-clients/AstralCore/Sources/AstralCore/Chrome/TurnSelection.swift
// Validates exact guidance revision selections for the current conversation.
// Outbound chat messages carry nonempty selections unchanged for server-side authorization.

import Foundation

public struct TurnSelection: Equatable, Sendable {
    public struct Agent: Equatable, Sendable {
        public let agentId: String
        public let revisionId: String
    }

    public struct Reference: Equatable, Sendable, Identifiable {
        public let id: String
        public let revision: UInt64
    }

    public let version: Int
    public let agent: Agent?
    public let skills: [Reference]
    public let notes: [Reference]
    public let json: JSONValue
    public var isEmpty: Bool { agent == nil && skills.isEmpty && notes.isEmpty }
    var isGuidanceSelection: Bool { agent == nil || continuityUUID4(agent?.agentId) != nil }

    public static let empty = TurnSelection(
        json: .object([
            "version": .number(1), "agent": .null, "skills": .array([]), "notes": .array([]),
        ]))!

    public init?(json: JSONValue?) {
        guard let json, let object = json.objectValue,
            Set(object.keys) == ["version", "agent", "skills", "notes"],
            object["version"]?.numberValue == 1,
            let rawAgent = object["agent"],
            let skills = Self.references(object["skills"], identityKey: "skill_id", maximum: 20),
            let notes = Self.references(object["notes"], identityKey: "note_id", maximum: 8)
        else { return nil }
        if rawAgent == .null {
            agent = nil
        } else {
            guard let fields = rawAgent.objectValue, Set(fields.keys) == ["agent_id", "revision_id"],
                let agentId = ConsoleDecoding.text(fields["agent_id"], maximum: 255),
                agentId == agentId.trimmingCharacters(in: .whitespacesAndNewlines),
                let revisionId = continuityUUID4(fields["revision_id"]?.stringValue)
            else { return nil }
            agent = Agent(agentId: agentId, revisionId: revisionId)
        }
        version = 1
        self.skills = skills
        self.notes = notes
        self.json = json
    }

    private static func references(_ json: JSONValue?, identityKey: String, maximum: Int) -> [Reference]? {
        guard let rows = ConsoleDecoding.rows(json, maximum: maximum) else { return nil }
        var result: [Reference] = []
        var identities: Set<String> = []
        for row in rows {
            guard let fields = row.objectValue, Set(fields.keys) == [identityKey, "revision"],
                let id = continuityUUID4(fields[identityKey]?.stringValue),
                let revision = ConsoleDecoding.number(fields["revision"], range: 1...9_007_199_254_740_991),
                revision.rounded() == revision, identities.insert(id).inserted
            else { return nil }
            result.append(Reference(id: id, revision: UInt64(revision)))
        }
        return result
    }
}

File: android-client/core/src/main/kotlin/com/personalailabs/astraldeep/core/chrome/TurnSelection.kt
// Retains exact agent, skill and note revisions selected for one conversation.
// Validated nonempty values are sent unchanged for server-side authorization.

package com.personalailabs.astraldeep.core.chrome

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.util.UUID

data class TurnSelection(
    val agent: Agent?,
    val skills: List<Reference>,
    val notes: List<Reference>,
    val json: JsonObject,
) {
    data class Agent(val agentId: String, val revisionId: String)

    data class Reference(val id: String, val revision: Long)

    val isEmpty: Boolean get() = agent == null && skills.isEmpty() && notes.isEmpty()
    val isGuidanceSelection: Boolean get() = agent?.let { runCatching { uuid(JsonPrimitive(it.agentId)) }.isSuccess } ?: true

    companion object {
        val EMPTY = requireNotNull(fromJson(Json.parseToJsonElement("""{"version":1,"agent":null,"skills":[],"notes":[]}""")))

        fun fromJson(value: JsonElement?): TurnSelection? =
            ConsoleDecoding.decode {
                val root = ConsoleDecoding.obj(value)
                require(root.keys == setOf("version", "agent", "skills", "notes"))
                require(ConsoleDecoding.number(root["version"], 1.0, 1.0) == 1.0)
                val agent =
                    if (root["agent"] == JsonNull) {
                        null
                    } else {
                        val fields = ConsoleDecoding.obj(root["agent"])
                        require(fields.keys == setOf("agent_id", "revision_id"))
                        val id = ConsoleDecoding.text(fields["agent_id"], 255)
                        require(id == id.trim())
                        Agent(id, uuid(fields["revision_id"]))
                    }
                TurnSelection(agent, references(root["skills"], "skill_id", 20), references(root["notes"], "note_id", 8), root)
            }

        private fun uuid(value: JsonElement?): String {
            val raw = ConsoleDecoding.text(value, 36)
            val uuid = UUID.fromString(raw)
            require(uuid.version() == 4 && uuid.variant() == 2 && uuid.toString() == raw)
            return raw
        }

        private fun references(
            value: JsonElement?,
            identity: String,
            maximum: Int,
        ): List<Reference> {
            val result =
                ConsoleDecoding.rows(value, maximum).map {
                    val row = ConsoleDecoding.obj(it)
                    require(row.keys == setOf(identity, "revision"))
                    val revision = ConsoleDecoding.number(row["revision"], 1.0, 9_007_199_254_740_991.0)
                    require(revision.toLong().toDouble() == revision)
                    Reference(uuid(row[identity]), revision.toLong())
                }
            require(result.map { it.id }.distinct().size == result.size)
            return result
        }
    }
}
